### ISSUE TITLE
entropy loss

### ISSUE THREAD

**Issue by apple55bc** [THUNLP-AIPoet/MixPoet#5]
layers.py line 480，Is there a mistake?
```python
entropy = torch.log(probs+1e-10) * probs # (B, n_class)
# should it be : ?
entropy = - torch.log(probs+1e-10) * probs # (B, n_class)
```

**Comment by XiaoyuanYi** [comment]
@apple55bc 

Hi, the definition of  entropy is  - sum_x  p(x) * log[p(x)].  However, we need to maximize entropy, please see Eq. (7) in our paper. When we use this entropy term as a kind of loss, please see line 154 in mix_trainer.py, we have to minimize negative entropy, that is, sum_x  p(x) * log[p(x)],  which is directly used as line 480 in layers.py just for convenience.

**Comment by apple55bc** [comment]
But Eq. (7)  seems to maximize -(-sum_x p(x) * log[p(x)]) ... Shouldn't we need minimum entropy to minimize uncertainty? 
So I still have some doubts, thanks!

**Comment by XiaoyuanYi** [comment]
@apple55bc 
Hi,
Eq. (7) is just a derived lower bound of log p(x|w). We need to maximize log p(x|w), that is, to maximize the first line of Eq. (7): E[-L] + H[q].  When we maximize the second term H[q],  since H[q] = - sum q * log q, then we can directly minimize sum q * log q. 

In Eq. (7), we need to maximize entropy rather than minimize it.  The entropy term can be regarded as a kind of regularization.  Larger H[q] means the probability q tends to be the uniform distribution, which can prevent q from being biased, for example,  prevent q(y|c) from assigning the most probability to only one single value y and giving other values nearly zero probability whatever the condition c is. 

We can also take log q in H[q] into the former expectation term E, then we can get another form of Eq. (7), which is just like Eq. (6): E[log p1] - KL[q||q1], then maximizing Eq. (7) is equivalent to maximizing the reconstruction probability log p1 and minimizing the KL divergence of the prior and posterior distributions.

**Comment by apple55bc** [comment]
I understand. Thank you very much!
